File: js/board.js
// לוח המשחק.
//
// המודול הבסיסי (עומק שורת תאים 5.00, מעבר 6.20, עמודים בעובי 30 ס"מ) חולץ
// מתוכנית חניון תת קרקעי מבוצע. כל סימוני החניה, המידות, טבלת השרטוט ושמות
// הפרויקט והמשרד המתכנן הוסרו.
//
// מרווחי העמודים גוונו בכוונה כדי לייצר מגוון מקרים: מרווח שמתאים לשלושה
// תאים, לזוג, ליחיד בלבד, ומרווח רחב שנראה כאילו נכנס בו עוד תא ולא נכנס.
// שני קירות ממ"ד יורדים לתוך הקומה וחוסמים חלק מהמרווח, כמו בחניון אמיתי.
//
// כל המידות במטרים.

console.log("[board] טוען קליפת חניון");

window.parkingBoard = {
  units: "m",
  width: 44.00,
  height: 16.70,
  stallLength: 5.00,
  aisle: { y0: 5.00, y1: 11.20, width: 6.20 },

  // עמודים בעובי 30 ס"מ. הכניסה לרמפה בקצה המערבי.
  columnDepth: 0.40,
  columnWidth: 0.30,

  rows: [
    {
      side: "lower", y0: 0.00, y1: 5.00, facing: "up",
      spans: [
        { x0: 0.60,  x1: 8.40,  clear: 7.80, left: "wall",   right: "column" },
        { x0: 8.70,  x1: 14.10, clear: 5.40, left: "column", right: "column" },
        { x0: 14.40, x1: 17.40, clear: 3.00, left: "column", right: "mamad" },
        { x0: 19.40, x1: 27.20, clear: 7.80, left: "mamad",  right: "column" },
        { x0: 27.50, x1: 33.70, clear: 6.20, left: "column", right: "column" },
        { x0: 34.00, x1: 39.40, clear: 5.40, left: "column", right: "column" },
      ],
    },
    {
      side: "upper", y0: 11.20, y1: 16.20, facing: "down",
      spans: [
        { x0: 2.20,  x1: 10.00, clear: 7.80, left: "column", right: "column" },
        { x0: 10.30, x1: 14.80, clear: 4.50, left: "column", right: "column" },
        { x0: 15.10, x1: 23.40, clear: 8.30, left: "column", right: "column" },
        { x0: 23.70, x1: 29.10, clear: 5.40, left: "column", right: "mamad" },
        { x0: 31.10, x1: 34.30, clear: 3.20, left: "mamad",  right: "column" },
        { x0: 34.60, x1: 42.40, clear: 7.80, left: "column", right: "wall" },
      ],
    },
  ],

  // קירות הממ"ד שיורדים לתוך הקומה. הם חוסמים את המרווח בין שני מפרצים.
  mamads: [
    { x0: 17.40, x1: 19.40, y0: 0.00, y1: 5.00, label: "ממ\"ד" },
    { x0: 29.10, x1: 31.10, y0: 11.20, y1: 16.20, label: "ממ\"ד" },
  ],

  // הרמפה, בקצה המערבי. לא ניתן לתכנן בה תאים.
  ramp: { x0: 0.00, x1: 0.60, y0: 5.00, y1: 11.20, label: "לרמפה" },
};

// עמודים נגזרים מקצות המרווחים, כדי שהשרטוט והחישוב לא ייפרדו לעולם.
(function deriveColumns(B) {
  const cols = [];
  B.rows.forEach((row) => {
    const yc = row.side === "upper" ? row.y0 : row.y1 - B.columnDepth;
    row.spans.forEach((sp) => {
      if (sp.left === "column") cols.push({ x0: sp.x0 - B.columnWidth, y0: yc, w: B.columnWidth, h: B.columnDepth });
      if (sp.right === "column") cols.push({ x0: sp.x1, y0: yc, w: B.columnWidth, h: B.columnDepth });
    });
  });
  // מסירים כפילויות של עמוד משותף לשני מפרצים סמוכים
  const seen = new Set();
  B.columns = cols.filter((c) => {
    const k = c.x0.toFixed(2) + "," + c.y0.toFixed(2);
    if (seen.has(k)) return false;
    seen.add(k);
    return true;
  });
})(window.parkingBoard);

console.log("[board] " +
  window.parkingBoard.rows.reduce((n, r) => n + r.spans.length, 0) + " מפרצים, " +
  window.parkingBoard.columns.length + " עמודים");

File: tools/board.py
# הופך את shell.json ללוח משחק מנורמל.
#
# העיקרון: הסטודנט לא מקבל את המפרצים מוכנים. הוא מקבל **קווי עמודים**, והמרווח
# הפנוי בין שני עמודים סמוכים הוא המפרץ שבו הוא מחליט כמה תאים להכניס.
# זו בדיוק ההחלטה שהטבלה בעמוד 294 מכריעה.
import json

d = json.load(open("shell.json", encoding="utf-8"))

x0 = min(min(b["x0"] for r in d["rows"] for b in r["bays"]),
         min(min(w[0], w[2]) for w in d["walls"]))
lower, upper = d["rows"][0], d["rows"][1]

# גבול תחתון: הקצה האחורי של שורת החניה התחתונה (גב אל גב עם השורה הבאה)
y_bottom = lower["y0"]
y_top = max(max(w[1], w[3]) for w in d["walls"])

SHIFT_X, SHIFT_Y = -x0, -y_bottom
mx = lambda v: round(v + SHIFT_X, 2)
my = lambda v: round(v + SHIFT_Y, 2)

# ---- עמודים, מקובצים לפי שורה
cols_by_row = {"lower": [], "upper": []}
mid_aisle = (d["aisle"]["y0"] + d["aisle"]["y1"]) / 2
for c in d["columns"]:
    side = "upper" if (c["y0"] + c["y1"]) / 2 > mid_aisle else "lower"
    cols_by_row[side].append(c)
for k in cols_by_row:
    cols_by_row[k].sort(key=lambda c: c["x0"])

# ---- מרווחים פנויים בין עמודים סמוכים = המפרצים לתכנון
def spans(cols):
    out = []
    for i in range(len(cols) - 1):
        a, b = cols[i], cols[i + 1]
        clear = round(b["x0"] - a["x1"], 2)
        if clear < 2.0:
            continue
        out.append({"x0": mx(a["x1"]), "x1": mx(b["x0"]), "clear": clear})
    return out

board = {
    "units": "m",
    "source": "קליפה מתוכנית חניון תת קרקעי מבוצע. סימוני החניה, המידות והזיהוי הוסרו.",
    "width": round(mx(max(max(w[0], w[2]) for w in d["walls"])), 2),
    "height": round(my(y_top), 2),
    "stallLength": 5.00,
    "aisle": {"y0": my(d["aisle"]["y0"]), "y1": my(d["aisle"]["y1"]), "width": d["aisle"]["width"]},
    "rows": [
        {"side": "lower", "y0": my(lower["y0"]), "y1": my(lower["y1"]),
         "facing": "up", "spans": spans(cols_by_row["lower"])},
        {"side": "upper", "y0": my(upper["y0"]), "y1": my(upper["y1"]),
         "facing": "down", "spans": spans(cols_by_row["upper"])},
    ],
    "columns": [{"x0": mx(c["x0"]), "y0": my(c["y0"]), "w": round(c["x1"] - c["x0"], 2),
                 "h": round(c["y1"] - c["y0"], 2)} for c in d["columns"]],
    "walls": [[mx(w[0]), my(w[1]), mx(w[2]), my(w[3])] for w in d["walls"]],
}

json.dump(board, open("board.json", "w", encoding="utf-8"), ensure_ascii=False, indent=1)
print("board %.2f x %.2f m" % (board["width"], board["height"]))
for r in board["rows"]:
    cl = [s["clear"] for s in r["spans"]]
    print("  %-6s y %.2f..%.2f   %d מרווחים: %s" % (r["side"], r["y0"], r["y1"], len(cl), cl))
print("  aisle %.2f  columns %d  walls %d" % (board["aisle"]["width"], len(board["columns"]), len(board["walls"])))
